async_downloads/cache.py

```diff
@@ -51,11 +51,11 @@ def save_download(download_key, iterable=None, file=None):
         return
     if iterable is not None:
         try:
-            with SpooledTemporaryFile(max_size=IN_MEMORY_MAX_SIZE_BYTES, mode='r+', newline="") as temp_file:
+            with SpooledTemporaryFile(max_size=IN_MEMORY_MAX_SIZE_BYTES, mode='w+', newline="", encoding="utf-8") as temp_file:
                 writer = csv.writer(temp_file, lineterminator="\n")
                 for row in iterable:
                     writer.writerow(row)
-                default_storage.save(download["filepath"], temp_file)
+                default_storage.save(download["filepath"], temp_file._file.buffer)
```

Comment:
Why do we have to do this?

What is going wrong with S3 backend?

Reply:
It does a MD5 hash at some point which fails because the file is opened as text rather than binary:
`Unicode-objects must be encoded before hashing`
Open to other suggestions of how to make this work - I'm definitely working at the edge of my understanding.

Reply:
(and we can't open the file as binary in the first place because csv writer needs to write to a file that's opened in a text mode)

Reply:
I'm wondering why we didn't encounter this problem before as well. I guess the difference is we are now opening a file whereas before it was a StringIO which we then wrapped in that `ContentFile(output.getvalue().encode())`, and the `encode()` turns it into bytes?

Reply:
It looks like `buffer` on `_file` only exists if the SpooledTemporaryFile has "rolled over" (gone from in-memory to disk).

```pycon
>>> with SpooledTemporaryFile(mode='w+', newline="") as f:
...     f.rollover()
...     print(f._file.buffer)
...     
<_io.BufferedRandom name=6>
>>> with SpooledTemporaryFile(mode='w+', newline="") as f:
...     print(f._file.buffer)
...     
Traceback (most recent call last):
  File "<input>", line 2, in <module>
AttributeError: '_io.StringIO' object has no attribute 'buffer'
```

Reply:
We can do the same thing we had before I think:
```pycon
with SpooledTemporaryFile(mode='w+', newline="") as f:
...     print(f._file.getvalue().encode())
...     
b''
```
But we should probably avoid accessing a "private" attribute `_file` (I don't know the API well enough to know if this is a "safe" one), and instead we should be able to use `read`:
```suggestion
                default_storage.save(download["filepath"], temp_file.read().buffer)
```

Reply:
Having said that, I guess that could cause exactly the same memory issues again...

Reply:
What version of python are you using?  When I test without rollover I get a `TextIOWrapper` not a `StringIO` object which works.

![image](https://user-images.githubusercontent.com/37864918/149140933-3ec2ae58-3f29-43c7-a918-a66ed71d4cd6.png)

Reply:
It was 3.8.

Reply:
And what does your SpooledTemporaryFile definition look like?

Reply:
Uses StringIO, but as discussed elsewhere, I am not on the latest 3.8.x, and this was a bug fixed on 3.9 and backported to 3.7 and 3.8.

https://github.com/python/cpython/commit/ea9835c5d154ab6a54eed627958473b6768b28cc
https://bugs.python.org/issue26730